integration/test/Test/LegalHold.hs

```diff
@@ -274,3 +280,522 @@ testLHDeleteClientManually = do
+        reqNotEnabled requester bob
+        statusShouldbe "no_consent"
+
+      legalholdWhitelistTeam tid alice >>= assertSuccess
+      postLegalHoldSettings tid alice (mkLegalHoldSettings lhPort) >>= assertSuccess
+
+      statusShouldbe "disabled"
+
+      requestLegalHoldDevice tid alice bob >>= assertStatus 201
+      statusShouldbe "pending"
+
+      -- requesting twice should be idempotent wrt the approval
+      -- mind that requesting twice means two "user.legalhold-request" notifications
+      -- for the clients of the user under legalhold (bob)
+      requestLegalHoldDevice tid alice bob >>= assertStatus 204
+      statusShouldbe "pending"
+
+      [bobc1, bobc2] <- replicateM 2 do
+        objId $ addClient bob def `bindResponse` getJSON 201
+      for_ [bobc1, bobc2] \client ->
+        awaitNotification bob client noValue isUserLegalholdRequestNotif >>= \notif -> do
+          notif %. "payload.0.last_prekey" `shouldMatch` lpk
+          notif %. "payload.0.id" `shouldMatch` objId bob
+
+-- | pops a channel until it finds an event that returns a 'Just'
+--   upon running the matcher function
+checkChan :: HasCallStack => Chan t -> (t -> App (Maybe a)) -> App a
+checkChan chan match = do
+  tSecs <- asks ((* 1_000_000) . timeOutSeconds)
+
+  maybe (assertFailure "checkChan: timed out") pure =<< timeout tSecs do
+    let go = readChan chan >>= match >>= maybe go pure
+    go
+
+-- | like 'checkChan' but throws away the request and decodes the body
+checkChanVal :: HasCallStack => Chan (t, LazyByteString) -> (Value -> MaybeT App a) -> App a
+checkChanVal chan match = checkChan chan \(_, bs) -> runMaybeT do
+  MaybeT (pure (decode bs)) >>= match
+
+setUpLHDevice ::
+  (HasCallStack, MakesValue tid, MakesValue owner, MakesValue uid) =>
+  tid ->
+  owner ->
+  uid ->
+  -- | the port the LH service is running on
+  Int ->
+  App ()
+setUpLHDevice tid alice bob lhPort = do
+  legalholdWhitelistTeam tid alice
+    >>= assertStatus 200
+
+  -- the status messages for these have already been tested
+  postLegalHoldSettings tid alice (mkLegalHoldSettings lhPort)
+    >>= assertStatus 201
+
+  requestLegalHoldDevice tid alice bob
+    >>= assertStatus 201
+
+  approveLegalHoldDevice tid bob defPassword
+    >>= assertStatus 200
+
+testLHApproveDevice :: App ()
+testLHApproveDevice = do
+  startDynamicBackends [mempty] \[dom] -> do
+    -- team users
+    -- alice (boss) and bob and charlie (member)
+    (alice, tid, [bob, charlie]) <- createTeam dom 3
+
+    -- ollie the outsider
+    ollie <- do
+      o <- randomUser dom def
+      connectTwoUsers o alice
+      pure o
+
+    -- sandy the stranger
+    sandy <- randomUser dom def
+
+    legalholdWhitelistTeam tid alice >>= assertStatus 200
+    -- TODO(mangoiv): it seems like correct behaviour to throw a 412
+    -- here, as we can only approve a device if we're in the pending
+    -- state. however, the old tests passed with a 403 which makes
+    -- this suspicious.
```

Comment:
```suggestion
```

(exceedingly weird, but there isn't really anything to do here, no?  you didn't change prod behavior here, so what worked before should keep working.)